### ISSUE TITLE
Need help with focus on the dynamic particular element.

### ISSUE THREAD

**Issue by pahonski** [dead/react-key-navigation#10]
![Снимок экрана 2019-03-19 в 10 03 05](https://user-images.githubusercontent.com/24761647/54586990-c6c18380-4a2f-11e9-944b-3e206307884a.png)
How can i put focus on the fourth element in the second vertical list programmatically.

**Comment by dead** [comment]
If you pass `forceFocus={true}` props to the fourth Focusable element it will be focused when the component mount.

**Comment by pahonski** [comment]
Thank you! But i create the sidebar component when my main component render. I do this because key-navigation doesn't work correct with many dynamic elements. I forced create all my focusable components when my main component render and set for them for example 'display: none'. My component mount 1 time and never unmounted while i am on the main component.

**Comment by dead** [comment]
Just fixed #3.
Maybe that will help you.

**Comment by pahonski** [comment]
How can i delete this elements (series-*) correct?
![Снимок экрана 2019-03-21 в 17 13 45](https://user-images.githubusercontent.com/24761647/54758329-07222c80-4bfd-11e9-81c8-bb853bfe1fc5.png)

I think this can help me with dynamic focusable elements. Now when i delete this components i get an error
![Снимок экрана 2019-03-21 в 17 17 22](https://user-images.githubusercontent.com/24761647/54758463-43ee2380-4bfd-11e9-8194-0e92d0fe889b.png)
but i deleted them before.

**Comment by pahonski** [comment]
When i used forseFocus props, it works.
![Снимок экрана 2019-03-21 в 18 00 22](https://user-images.githubusercontent.com/24761647/54761529-523f3e00-4c03-11e9-9953-1478e573047c.png)
But when i clicked UP, focus went on the first element, not on the third.

**Comment by pahonski** [comment]
Sorry, i don't understand how it works.

```
<VerticalList>
{this.state.dynamicInsert && <Search text="inserted!!" />}
</VerticalList>
```
In example which i found in the issue#3 it doesn't work.

**Comment by dead** [comment]
Did you update to 0.0.10? The same example in #3 with version 0.0.10 https://codesandbox.io/s/50xy2z3244

```jsx
<VerticalList>
{this.state.dynamicInsert && <Search text="inserted!!" />}
</VerticalList>
```
The code above works in version 0.0.9 because the "VerticalList" is already monted, so when the dynamic  component is inserted it will be in the correct order/position (the first element of that VerticalList). [Same example in #3 using the code above](https://codesandbox.io/s/znlxwk77rl)

In version 0.0.10 you don't need VerticalList because it will re-order the children components when the parent component is updated. So just using `{this.state.dynamicInsert && <Search text="inserted!!" />}` should be fine.

**Comment by pahonski** [comment]
But i need one horizontal list which include 2 vertical lists. The first list for seasons and the second for the series. When i remove Vertical list it doesn't work correct.
![Снимок экрана 2019-03-21 в 20 25 55](https://user-images.githubusercontent.com/24761647/54772028-963c3e00-4c17-11e9-90d3-00d5c4a7ed75.png)

And i get new error with version 0.0.10 when i re-render component withe series.
![Снимок экрана 2019-03-21 в 20 19 15](https://user-images.githubusercontent.com/24761647/54771572-a0117180-4c16-11e9-804b-79a380604623.png)
Could you tell me what will i need change in my code. I tried to do as you wrote, but it doesn't help me.
May be i can't use array with dynamic elements?

**Comment by pahonski** [comment]
https://youtu.be/bSeYduApEc4
I attached video with problem.

**Comment by dead** [comment]
I think I found out what was your problem, I'm going to fix it soon.

**Comment by pahonski** [comment]
I'm looking forward to fix this problem.

**Comment by dead** [comment]
Please try the v0.0.11

**Comment by pahonski** [comment]
Hello, i have new problem with focusable elements after update. I try to explain it soon. But old problem with correct KEY UP and KEY DOWN isn't fixed.

**Comment by pahonski** [comment]
https://youtu.be/mCmKylez4s4

Now i can't switch to window with PLAY button when i press KEY_UP. 
![Снимок экрана 2019-03-24 в 10 25 53](https://user-images.githubusercontent.com/24761647/54876350-8abb6380-4e1f-11e9-8d9e-c3c74140b3b9.png)

This element doesn't dynamic. It has diplay:none. When data are received element is changed display to flex.
In the version 0.0.10 it works good, but now it doesn't work.

Main component with navigation:
![Снимок экрана 2019-03-24 в 10 33 22](https://user-images.githubusercontent.com/24761647/54876398-598f6300-4e20-11e9-8ff6-49f0ad9900f0.png)

Menu component
![Снимок экрана 2019-03-24 в 10 34 32](https://user-images.githubusercontent.com/24761647/54876421-b559ec00-4e20-11e9-8c15-68ea373cd7d1.png)

ContentList component (films posters).
![Снимок экрана 2019-03-24 в 10 36 04](https://user-images.githubusercontent.com/24761647/54876429-d7ec0500-4e20-11e9-9f49-62c681fb596e.png)

I don't know if i explained well. I will record video if you don't understand my problem.

**Comment by pahonski** [comment]
https://youtu.be/t7VmV8o5dCo

In this video i compare version 9 and 11

**Comment by dead** [comment]
Thanks for the feedback! I guess your problem now is because

**Comment by pahonski** [comment]
I noticed that the focus is lost when the component is updated. In my case it happened when i get new data from redux. It starts to work incorrectly.

**Comment by dead** [comment]
Thanks for the feedback! But I couldn't reproduce your error...
Can you change to mount the component only when it should be visible?
That is the only thing that could bug the navigation system because even though it is not visible the navigation is not aware of that. If that is your problem I can create a props `disabled` that would make the navigation ignore it.

**Comment by pahonski** [comment]
Thank you, this props will be good. On this week i'll try to find, where i get this problem. I'll write about it soon.

**Comment by pahonski** [comment]
Now i describe other problems, which i noticed.
1. When my project was downloaded i see some react-dom errors in the console. Without last warning.
![1](https://user-images.githubusercontent.com/24761647/54976644-0b0ecf80-4fac-11e9-9e1f-892b728b76a1.jpg)
![2](https://user-images.githubusercontent.com/24761647/54976645-0b0ecf80-4fac-11e9-838b-6b75b6c73ee8.jpg)
2. I have search keyboard with focusable buttons, but it doesn't that behaviour which i wanted.
https://youtu.be/VAT4_FTInoE

When I focus keyboard and press KEYDOWN i want to go on the second row not ont the menu below.
This is the keyboard code:
![01](https://user-images.githubusercontent.com/24761647/54977579-d7817480-4fae-11e9-9e5c-fe0791434e8e.jpg)

This is the Menu code which include keyboard
![02](https://user-images.githubusercontent.com/24761647/54977608-ef58f880-4fae-11e9-95a8-c64e4b9cfaf5.jpg)

**Comment by pahonski** [comment]
3. Bug with dynamic element when it appears.
https://youtu.be/ibXfLau5IL4
this element
![6](https://user-images.githubusercontent.com/24761647/54979937-85901d00-4fb5-11e9-8307-0804cd1f536a.jpg)

**Comment by pahonski** [comment]
I tried dynamical elements in more simple component, but it doesn't work too. 

I have Tv component with navigation and other components.
![1](https://user-images.githubusercontent.com/24761647/54994411-407be300-4fd5-11e9-8db1-344e1307c6ef.jpg)
Sidebar same as in your example in youtube-tv.
![666](https://user-images.githubusercontent.com/24761647/54995005-c8aeb800-4fd6-11e9-8969-51a9c987abf7.jpg)
List with tv channels which i get from ajax
![3](https://user-images.githubusercontent.com/24761647/54994557-99e41200-4fd5-11e9-8e26-ba109c5d65d3.jpg)
And list items
![4](https://user-images.githubusercontent.com/24761647/54994620-c009b200-4fd5-11e9-862a-d7492ed12480.jpg)

Focus works only on sidebar static element, when i get tv channels list from ajax, they don't get focus.
video
https://youtu.be/l4VPEaWwYGg

**Comment by pahonski** [comment]
I added code on the codesandbox.

https://codesandbox.io/s/xlnvr9zr9w

**Comment by dead** [comment]
Update to v.0.0.12 https://codesandbox.io/s/622xk79vvw and now this example work 😄

**Comment by pahonski** [comment]
Hello i noticed another problem. 
When i render Focusable items as in this example this code works good. 
![1](https://user-images.githubusercontent.com/24761647/55061977-11717a00-5086-11e9-9a63-3b138ae08ff1.jpg)

But when i render Focusable element in the component as in this example it works one time. Sidebar open and close. But then i don't focus this sidebar.
![3](https://user-images.githubusercontent.com/24761647/55062140-601f1400-5086-11e9-8447-4b235980629c.jpg)

This is component which include focusable.
![2](https://user-images.githubusercontent.com/24761647/55062158-6d3c0300-5086-11e9-96cb-d58205040e59.jpg)

**Comment by dead** [comment]
Can you reproduce this in codesandbox? I tried https://codesandbox.io/s/ryrnxz902n but it works

**Comment by pahonski** [comment]
Maybe this video can help.
https://youtu.be/nHktSpwZ168

**Comment by pahonski** [comment]
In your case it doesn't work too, https://codesandbox.io/s/ryrnxz902n.

**Comment by pahonski** [comment]
I found this bug.
https://codesandbox.io/s/72o5yj4mn1

**Comment by dead** [comment]
Can you try the new version (v0.0.13)?

**Comment by pahonski** [comment]
Yes, it's working. Thank you. Did you add props "disabled"? And how can i set navDeafult on the vertical list which include dynamic elements?

**Comment by dead** [comment]
Still haven't added the `disabled` props. The `navDeafult` sets the element the navigation will focus when it will lose a focus or the when the navigation start (after loading the page). You want something that when you enter a VerticalList it will focus a specific element?

**Comment by pahonski** [comment]
It would be nice:
1) when I enter a VerticalList it will focus a specific element.
2) when i have opportunity to set navDefault on the VerticalList which doesn't has children elements, but it is going to get them from ajax.

**Comment by pahonski** [comment]
And props "disabled" :)

**Comment by pahonski** [comment]
Hello, are you going to make changes soon?

**Comment by pahonski** [comment]
How can i clean this list http://joxi.ru/brRgav6i7Qaeam.
When i create a lot of dynamic elements this list gets bigger and bigger.

**Comment by dead** [comment]
Hey, I kind busy this week, so I'm not sure if i will able to do something.

Does this list cause you performance issues?
I only use that to keep a reference of the focusable components.

**Comment by pahonski** [comment]
What about this? It is very important.
http://joxi.ru/VrwyVZlI7OKjO2

About performance. Now i don't have problems but in future it might be.

**Comment by pahonski** [comment]
Hello, i have a new problem with forceFocus.
In this app i show a simple example.
https://codesandbox.io/s/72o5yj4mn1

I tried forceFocus from left sidebar. I have to focus right items when I press SPACE button. It works, but then when i return on the left sidebar after this manipulations, KEY_UP and KEY_DOWN doesn't work.

When i have to change SPACE (which i described above) on KEY_LEFT? Can I use event.stopPropagation and forceFocus?

**Comment by pahonski** [comment]
Could you fix at least this bug please?

**Comment by dead** [comment]
I going to try fixing this issue tomorrow.

**Comment by pahonski** [comment]
Thank you. I will looking forward to.

**Comment by dead** [comment]
Sorry, I'm not going to be able to fix that today. I'm without internet at home. Hope I can fix it Monday.

**Comment by pahonski** [comment]
Sure. I hope it too.

**Comment by pahonski** [comment]
Unfortunately forceFocuses don't work anywhere :( I noticed that things which worked before, don't work now.

**Comment by dead** [comment]
Yeah, I tried to fix the issue yesterday, but no success. The problem all comes down to the issue #3. The "solution" I found is wrong and doesn't work properly at all.
Probably I will need to rewrite most part of the library to work around this (and not even sure if this is possible).

I recommend you to try https://github.com/dead/react-js-spatial-navigation or even the pure js library https://github.com/luke-chang/js-spatial-navigation

**Comment by pahonski** [comment]
We wrote much code with your library. Could you finish some important things for a fee?

**Comment by pahonski** [comment]
Thank you for a help. I'll try your library  https://github.com/dead/react-js-spatial-navigation. But i don't see in it forceFocus :).

**Comment by pahonski** [comment]
And when i updated to version 0.0.5 your example doesn't work correct.
